allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+
+
+@CrossValidator.register("predefined_split")
+class PredefinedSplit(CrossValidator, sklearn.model_selection.PredefinedSplit):
+    def __init__(self, test_fold: Sequence[int]) -> None:
+        super().__init__(test_fold=test_fold)
+
+
+@CrossValidator.register("repeated_k_fold")
+class RepeatedKFold(CrossValidator, sklearn.model_selection.RepeatedKFold):
+    def __init__(
+        self,
+        n_splits: int = 5,
+        n_repeats: int = 10,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(n_splits=n_splits, n_repeats=n_repeats, random_state=random_state)
+
+
+@CrossValidator.register("repeated_stratified_k_fold")
+class RepeatedStratifiedKFold(CrossValidator, sklearn.model_selection.RepeatedStratifiedKFold):
+    def __init__(
+        self,
+        n_splits: int = 5,
+        n_repeats: int = 10,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(n_splits=n_splits, n_repeats=n_repeats, random_state=random_state)
+
+
+@CrossValidator.register("shuffle_split")
+class ShuffleSplit(CrossValidator, sklearn.model_selection.ShuffleSplit):
+    def __init__(
+        self,
+        n_splits: int = 10,
+        test_size: Optional[Union[float, int]] = None,
+        train_size: Optional[Union[float, int]] = None,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(
+            n_splits=n_splits, test_size=test_size, train_size=train_size, random_state=random_state
+        )
+
+
+@CrossValidator.register("stratified_k_fold")
+class StratifiedKFold(CrossValidator, sklearn.model_selection.StratifiedKFold):
+    def __init__(
+        self,
+        n_splits: int = 5,
+        shuffle: bool = False,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(n_splits=n_splits, shuffle=shuffle, random_state=random_state)
+
+
+@CrossValidator.register("stratified_shuffle_split")
+class StratifiedShuffleSplit(CrossValidator, sklearn.model_selection.StratifiedShuffleSplit):
+    def __init__(
+        self,
+        n_splits: int = 10,
+        test_size: Optional[Union[float, int]] = None,
+        train_size: Optional[Union[float, int]] = None,
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(
+            n_splits=n_splits, test_size=test_size, train_size=train_size, random_state=random_state
+        )
+
+
+@CrossValidator.register("time_series_split")
+class TimeSeriesSplit(CrossValidator, sklearn.model_selection.TimeSeriesSplit):
+    def __init__(self, n_splits: int = 5, max_train_size: Optional[int] = None) -> None:
+        super().__init__(n_splits=n_splits, max_train_size=max_train_size)
+
+
+logger = logging.getLogger(__name__)
+
+
+@TrainModel.register("cross_validation", constructor="from_partial_objects")
+class CrossValidateModel(TrainModel):
```

Comment:
I don't think I need to subclass `TrainModel` but it felt like the best thing to do though. Maybe there should be an abstract class for both (and the other one maybe name it `HoldoutTrain`?), which just defines the contract?

Reply:
If we do this, the base class should be `Command`.  That's basically what's going on here; we're defining a command and a config file specification.  Doing this should let us remove a bunch of boilerplate in our command code.

Reply:
I'd be happy to look at that as a separate PR, but let's not do it here.

Reply:
You mean to refactor out `TrainModel` behavior first in another PR, then come back here and rebase it to continue here?

Reply:
`Command` sounds confusing when also having `Subcommand`. What about `TrainingMethod` (because it's used only with `allennlp train`).

Reply:
No, I mean let's not invent new infrastructure in this PR.

Let's also keep in mind that there are only two cases here, training and cross-validation. We don't need to invent a whole new infrastructure as if we'd have dozens of similar training procedures. There are just two, which (maybe) need to share a bit of code.

It was my understanding that you would put this into a `Subcommand`. Did I misunderstand that?

Reply:
I definitely agree with not doing anything like this in this PR.  Sorry, that wasn't what I was suggesting at all.

In this PR, the base class for `CrossValidateModel` should be `Registrable`, and it should be structured similarly to `TrainModel`.  The class should live in `commands`, because that's what it is.

*If* we want an abstract class for both, it should be something like `Command`.  That's a totally separate issue that would need a bunch of design work to get it right, and may or may not be worth the effort.

Reply:
The abstraction is already there by subclassing (and using `run` and `finish`). It's just making it cleaner by making an interface out of it. What if I do it here and we see what do we do?

Reply:
No, subclassing is definitely not the right thing to do.  You can implement the same methods, if you want, to make it easier to make an abstraction out of this in the future, but inheritance is the wrong relationship between these two objects.

Reply:
In general: inheritance is useful for polymorphism.  We make use of that quite a bit in allennlp.  But that's about the only place that inheritance should be used.  If you are trying to share code outside of polymorphism, there is basically always a better option.

Reply:
Yeah, I completely agree, just describing how it is now.

Reply:
See 52f0387

Reply:
Please don't add an abstraction in this PR.  Just create a `CrossValidateModel(Registrable)` class.  What abstraction, if any, should subsume those two commands needs a bunch of design thought, and we shouldn't just introduce something here that we very well might want to remove later.